cloud/services/compute/loadbalancers/reconcile.go

```diff
@@ -18,72 +18,230 @@ package loadbalancers
 
 import (
 	"context"
+	"errors"
+	"strings"
 
 	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/meta"
 	"google.golang.org/api/compute/v1"
-
 	"k8s.io/utils/ptr"
+	infrav1 "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-gcp/cloud/gcperrors"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-// Reconcile reconcile cluster control-plane loadbalancer compoenents.
+// loadBalancingMode describes the load balancing mode that the backend performs.
+type loadBalancingMode string
+
+const (
+	// Utilization determines how the traffic load is spread based on the
+	// utilization of instances.
+	loadBalancingModeUtilization = loadBalancingMode("UTILIZATION")
+
+	// Connection determines how the traffic load is spread based on the
+	// total number of connections that a backend can handle. This is
+	// only mode available for passthrough Load Balancers.
+	loadBalancingModeConnection = loadBalancingMode("CONNECTION")
+
+	loadBalanceTrafficInternal = "INTERNAL"
+)
+
+// Reconcile reconcile cluster control-plane loadbalancer components.
 func (s *Service) Reconcile(ctx context.Context) error {
 	log := log.FromContext(ctx)
 	log.Info("Reconciling loadbalancer resources")
+
+	// Creates instance groups used by load balancer(s)
 	instancegroups, err := s.createOrGetInstanceGroups(ctx)
 	if err != nil {
 		return err
 	}
 
-	healthcheck, err := s.createOrGetHealthCheck(ctx)
+	lbSpec := s.scope.LoadBalancer()
+	lbType := ptr.Deref(lbSpec.LoadBalancerType, infrav1.ExternalLoadBalancerOnly)
+	// Create a Global External Proxy Load Balancer by default
+	if lbType == infrav1.ExternalLoadBalancerOnly || lbType == infrav1.DualLoadBalancer {
+		if err = s.createExternalLoadBalancer(ctx, lbType, instancegroups); err != nil {
+			return err
+		}
+	}
+
+	// Create a Regional Internal Passthrough Load Balancer if configured
+	if lbType == infrav1.DualLoadBalancer || lbType == infrav1.InternalLoadBalancerOnly {
+		name := infrav1.InternalRoleTagValue
+		if lbSpec.InternalLoadBalancer != nil {
+			name = ptr.Deref(lbSpec.InternalLoadBalancer.Name, infrav1.InternalRoleTagValue)
+		}
+		if err = s.createInternalLoadBalancer(ctx, name, lbType, instancegroups); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Delete delete cluster control-plane loadbalancer components.
+func (s *Service) Delete(ctx context.Context) error {
+	lbSpec := s.scope.LoadBalancer()
+	lbType := ptr.Deref(lbSpec.LoadBalancerType, infrav1.ExternalLoadBalancerOnly)
+	if lbType == infrav1.ExternalLoadBalancerOnly || lbType == infrav1.DualLoadBalancer {
+		if err := s.deleteExternalLoadBalancer(ctx); err != nil {
+			return err
```

Comment:
Should we attempt to also delete Internal Load Balancer when lbType == infrav1.DualLoadBalancer and this delete fails?